### PR TITLE
Remove runtime imports in realize

### PR DESCRIPTION
**Pull request by ryandorrington**
Conditional statements have been relocated from _realize_from to the buffers, allowing for the removal of runtime imports.

Partly solves #1653?

cleans up realize.py but adds complexity to lib.py. Room for improvement?

### CODE REVIEW

File: tinygrad/realize.py [hunk 1]
```diff
@@ -49,16 +47,7 @@ def _realize_from(buffer: LazyBuffer, src: LazyBuffer) -> None:
   assert src.realized.size == buffer.st.size(), f"size mismatch on FROM {src.realized.size} != {buffer.st.size()}"
   assert src.st.contiguous and buffer.st.contiguous, "all must be contiguous for from"
   if DEBUG >= 2: print(f"***      copy {buffer.device} <- {src.device} size {src.realized.size:<16d} shape {str(buffer.shape):23s} dtype {src.realized.dtype}")
-  # TODO: make this generic
-  if isinstance(src.realized, RawDiskBuffer) and issubclass(Device[buffer.device].buffer, RawBufferMapped):
```
Comment (geohot) [inline tinygrad/realize.py:53]: Hmm, can DiskBuffer be implemented as transfer?

File: tinygrad/runtime/lib.py [hunk 2]
```diff
@@ -41,6 +43,14 @@ def _buffer(self) -> memoryview: raise NotImplementedError("must be implemented"
   def toCPU(self) -> np.ndarray: return np.frombuffer(self._buffer(), dtype=np.dtype(self.dtype.np, metadata={"backing": self}), count=self.size)  # type: ignore
   def _copyin(self, x:np.ndarray) -> None: np.copyto(self.toCPU(), x.reshape(-1))
 
+  @classmethod
+  def fromBuffer(cls, src, shape, dtype, **kwargs):
+    from tinygrad.runtime.ops_disk import RawDiskBuffer
+    if isinstance(src.realized, RawDiskBuffer):
+      assert all_int(shape), "does not support symbolic shape"
+      src.realized.readinto(cast(RawBufferMapped, cls(prod(shape), dtype, **kwargs))._buffer())
```
Comment (geohot) [inline tinygrad/runtime/lib.py:51]: Like this shouldn't be here, it should be in RawDiskBuffer

Comment (ryandorrington) [reply tinygrad/runtime/lib.py]: @geohot, could you please elaborate? Are you suggesting adding a 'transfer' method to RawDiskBuffer (as currently done), reimplementing RawDiskBuffer as a subclass of RawBufferTransfer, or something different?

File: tinygrad/runtime/lib.py [hunk 3]
```diff
@@ -42,6 +44,13 @@ def buffer_view(self) -> np.ndarray: return np.frombuffer(self._buffer(), dtype=
   def toCPU(self) -> np.ndarray: return self.buffer_view().copy() # Need a copy, since jit will write to the same buffer.
   def _copyin(self, x:np.ndarray) -> None: np.copyto(self.buffer_view(), x.reshape(-1))
 
+  @classmethod
+  def fromBuffer(cls, src, shape, dtype, **kwargs):
+    from tinygrad.runtime.ops_disk import RawDiskBuffer
+    if isinstance(src.realized, RawDiskBuffer):
```
Comment (geohot) [inline tinygrad/runtime/lib.py:50]: Can RawDiskBuffer be RawBufferTransfer and this code can be shared?

**Pull request merged**